[scroll-animations] Force render tree creation under TimelineRange::cssToLengthConversionData

nmoucht opened the pull request:
#### e535237203c31345e9c2eabc0869d3b1c69aa8e7
<pre>
[scroll-animations] Force render tree creation under TimelineRange::cssToLengthConversionData
<a href="https://bugs.webkit.org/show_bug.cgi?id=282708">https://bugs.webkit.org/show_bug.cgi?id=282708</a>
<a href="https://rdar.apple.com/139380230">rdar://139380230</a>

Reviewed by NOBODY (OOPS!).

Force render tree creation under TimelineRange::cssToLengthConversionData as we need
conversion data when converting provided animation-range via the JS API. This fixes the
demo <a href="https://scroll-driven-animations.style/demos/cover-to-fixed-header/waapi/">https://scroll-driven-animations.style/demos/cover-to-fixed-header/waapi/</a> where we
flakily wouldn&apos;t play the animation if there was no render tree created when provided the
animation-range.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-timeline-range-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-range-expected.txt:
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::cssToLengthConversionData):
(WebCore::SingleTimelineRange::lengthForCSSValue):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e535237203c31345e9c2eabc0869d3b1c69aa8e7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81281 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27989 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78830 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4041 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60134 "Build is in progress. Recent messages:") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18216 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79780 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50056 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40456 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23363 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26313 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68576 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23693 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82690 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2698 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68401 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4242 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67653 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11626 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9715 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4036 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6845 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4059 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7489 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5817 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -65,11 +65,13 @@ TimelineRangeValue SingleTimelineRange::serialize() const
 static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(RefPtr<Element> element)
 {
+    Ref document = element->document();
+    document->updateLayoutIgnorePendingStylesheets();
```
I think this is too deep. What's the entry point from JS that we get here by?

nmoucht replied:
This is called via setBindingsRangeStart/End via the WebAnimations JS API. Moved this up into SingleTimelineRange::parse which is directly called by those setters.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -85,13 +86,17 @@ static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(
     if (value->valueID() == CSSValueAuto)
         return { };
 
+    element->document().updateLayoutIgnorePendingStylesheets();
```
`protectedDocument()`

graouts replied:
What happened to moving this up to `SingleTimelineRange::parse()`?

nmoucht replied:
Fixed. Moving it farther up caused a test failure.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -141,6 +146,7 @@ SingleTimelineRange SingleTimelineRange::parse(TimelineRangeValue&& value, RefPt
     RefPtr document = element->protectedDocument();
     if (!document)
         return { };
+
```
This needs not be added.

nmoucht replied:
Removed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -65,11 +65,12 @@ TimelineRangeValue SingleTimelineRange::serialize() const
 static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(RefPtr<Element> element)
 {
+    Ref document = element->document();
+
```
No need for this empty line.

nmoucht replied:
Fixed.

graouts reviewed (changes requested):
ChangeLog is outdated.

nmoucht committed:
[scroll-animations] Force render tree creation under TimelineRange::cssToLengthConversionData
https://bugs.webkit.org/show_bug.cgi?id=282708
rdar://139380230

Reviewed by NOBODY (OOPS!).

Force render tree creation under TimelineRange::cssToLengthConversionData as we need
conversion data when converting provided animation-range via the JS API. This fixes the
demo https://scroll-driven-animations.style/demos/cover-to-fixed-header/waapi/ where we
flakily wouldn't play the animation if there was no render tree created when provided the
animation-range.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-timeline-range-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-range-expected.txt:
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::cssToLengthConversionData):
(WebCore::SingleTimelineRange::lengthForCSSValue):

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -86,13 +86,17 @@ static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(
     if (value->valueID() == CSSValueAuto)
         return { };
 
+    element->protectedDocument()->updateLayoutIgnorePendingStylesheets();
```
@smfr do you think this remains too deep?
